### PR TITLE
Issue #528: Record SparkJob metrics

### PR DESCRIPTION
**Pull request by robhudson**

### CODE REVIEW

File: atmo/jobs/models.py [hunk 1]
```diff
@@ -406,17 +407,54 @@ def sync(self, info=None):
...
+            if self.status == Cluster.STATUS_TERMINATED_WITH_ERRORS:
+                transaction.on_commit(lambda: self.alert(info))
+            self.save()
+
+        if date_fields_updated:
+            if self.finished_at:
+                # When job is finished, record normalized instance hours.
+                hours = math.ceil(
+                    (self.finished_at - self.started_at).seconds / 60 / 60
+                )
```
Comment (jezdez) [inline atmo/jobs/models.py]: This looks like a copy/paste of the one in clusters app, can we refactor this into one method please?

File: atmo/jobs/models.py [hunk 2]
```diff
@@ -406,17 +407,54 @@ def sync(self, info=None):
...
+        date_fields_updated = False
+
         # set the various model fields to the value the API returned
         for api_field, model_field in model_field_map:
             field_value = info.get(api_field)
-            if field_value is None:
+            if field_value is None or field_value == getattr(self, model_field):
                 continue
             setattr(self, model_field, field_value)
+            save_needed = True
```
Comment (jezdez) [inline atmo/jobs/models.py]: This is basically a rewrite of the ["update_fields" feature](https://docs.djangoproject.com/en/1.11/ref/models/instances/#specifying-which-fields-to-save) in Django, let's use this instead, no need to track the state ourselves.

File: atmo/jobs/models.py [hunk 3]
```diff
@@ -406,17 +407,54 @@ def sync(self, info=None):
...
+                # scheduled job to run.
+                run_time = (self.finished_at - self.ready_at).seconds
+                Metric.record(
+                    'sparkjob-run-time', run_time,
+                    data={
+                        'identifier': self.spark_job.identifier,
+                        'size': self.size,
+                        'jobflow_id': self.jobflow_id,
+                    }
+                )
```
Comment (jezdez) [inline atmo/jobs/models.py]: The record methods uses the model manager `create` method, which means it won't be using an own transaction block and could cause a race condition. Let's wrap this either into an `transaction.on_commit` call or the record method in a `transaction.atomic()` block. The latter is easier since it'll apply to all calls to `Metric.record` but it'll require careful consideration with the app code where it's called from to prevent creating a metric before the app code has been processed. So a `transaction.on_commit` is preferable.

File: tests/jobs/test_models.py [hunk 4]
```diff
@@ -146,7 +147,7 @@ def test_sync_terminated(request, mocker, sparkjob_provisioner_mocks,
         'atmo.clusters.provisioners.ClusterProvisioner.info',
         return_value={
             'creation_datetime': one_hour_ago,
-            'ready_datetime': None,
+            'ready_datetime': one_hour_ago,
```
Comment (jezdez) [inline tests/jobs/test_models.py]: Huh, why this change?

File: tests/jobs/test_models.py [hunk 5]
```diff
@@ -160,6 +161,21 @@ def test_sync_terminated(request, mocker, sparkjob_provisioner_mocks,
     assert spark_job.latest_run.finished_at == now
     assert not spark_job.latest_run.alerts.exists()
 
+    metric = Metric.objects.get(key='sparkjob-normalized-instance-hours')
```
Comment (jezdez) [inline tests/jobs/test_models.py]: Let's check here that there is exactly one `Metric` with that name and not use `get` to fetch it.

File: tests/jobs/test_models.py [hunk 6]
```diff
@@ -160,6 +161,21 @@ def test_sync_terminated(request, mocker, sparkjob_provisioner_mocks,
...
+    assert metric.value == 5
+    assert metric.data == {
+        'identifier': spark_job.identifier,
+        'size': spark_job.size,
+        'jobflow_id': spark_job.latest_run.jobflow_id,
+    }
+    metric = Metric.objects.get(key='sparkjob-run-time')
```
Comment (jezdez) [inline tests/jobs/test_models.py]: Same as above.

File: tests/jobs/test_models.py [hunk 7]
```diff
@@ -251,6 +267,12 @@ def factory():
     return factory
 
 
+def test_is_public(spark_job):
+    assert not spark_job.is_public
+    spark_job.result_visibility = models.SparkJob.RESULT_PUBLIC
+    assert spark_job.is_public
```
Comment (jezdez) [inline tests/jobs/test_models.py]: Yay, coverage!

File: atmo/jobs/models.py (hunk 2 above)
Comment (robhudson) [reply atmo/jobs/models.py]: What I'm trying to do is avoid a `save()` if the data didn't change. I noticed that calling `save()` on every sync call made the front-end trigger a UI refresh notice when no actual change actually occurs.

Does using `update_fields` bypass the auto-now stuff for the modified column?

File: atmo/jobs/models.py (hunk 1 above)
Comment (robhudson) [reply atmo/jobs/models.py]: Yeah, true. I actually typed it by using the cluster one as reference. I should have noticed myself how exact they were.

File: tests/jobs/test_models.py (hunk 4 above)
Comment (robhudson) [reply tests/jobs/test_models.py]: This is for the test for recording the "sparkjob-run-time" which calculates run time by using `end_datetime` minus `ready_datetime`. Also, it doesn't make sense to have a end date w/o a ready date.

File: tests/jobs/test_models.py (hunk 5 above)
Comment (robhudson) [reply tests/jobs/test_models.py]: Curious why? The `get` would fail if there are not exactly 1.

File: tests/jobs/test_models.py (hunk 7 above)
Comment (robhudson) [reply tests/jobs/test_models.py]: I was checking the coverage on jobs/models.py and noticed this was an easy one to throw in. :)

**Comment by robhudson** [comment]
Updated to wrap metrics recording in transactions.

Any further refactoring will be done in later commits.

**Comment by robhudson** [comment]
Tested locally with clusters and sparkjobs and looks good.

File: tests/jobs/test_models.py (hunk 4 above)
Comment (jezdez) [reply tests/jobs/test_models.py]: Got it.

File: tests/jobs/test_models.py (hunk 5 above)
Comment (jezdez) [reply tests/jobs/test_models.py]: Right, it's just a nicer assertion than having to decrypt why it would be returning a `MultipleObjectsReturned` exception

File: tests/jobs/test_models.py (hunk 7 above)
Comment (jezdez) [reply tests/jobs/test_models.py]: Cool!
